kubernetes/cache.go

```diff
@@ -0,0 +1,458 @@
+		maxSyncCount    int
+		isErrorState    bool
+		lastError       error
+		lastErrorLock   sync.Mutex
+		controllers     map[string]cache.SharedIndexInformer
+	}
+)
+
+var (
+	lastCacheErrorLock sync.Mutex
+	errorCallbacks     []func(error)
+)
+
+func init() {
+	setupErrorHandlers()
+	errorCallbacks = make([]func(error), 0)
+}
+
+func setupErrorHandlers() {
+	nErrFunc := len(utilruntime.ErrorHandlers)
+	customErrorHandler := make([]func(error), nErrFunc+1)
+	for i, errorFunc := range utilruntime.ErrorHandlers {
+		customErrorHandler[i] = errorFunc
+	}
+	customErrorHandler[nErrFunc] = func(err error) {
+		for _, callback := range errorCallbacks {
+			callback(err)
+		}
+	}
+	utilruntime.ErrorHandlers = customErrorHandler
+}
+
+func registerErrorCallback(callback func(error)) {
+	defer lastCacheErrorLock.Unlock()
+	lastCacheErrorLock.Lock()
+	errorCallbacks = append(errorCallbacks, callback)
+}
+
+func newCacheController(clientset kube.Interface, refreshDuration time.Duration) cacheController {
+	newControllerImpl := controllerImpl{
+		clientset:       clientset,
+		refreshDuration: refreshDuration,
+		stopChan:        nil,
+		controllers:     initControllers(clientset, refreshDuration),
+		syncCount:       0,
+		maxSyncCount:    20, // Move this to config ? or this constant is good enough ?
+	}
+	registerErrorCallback(newControllerImpl.ErrorCallback)
+
+	return &newControllerImpl
+}
+
+func initControllers(clientset kube.Interface, refreshDuration time.Duration) map[string]cache.SharedIndexInformer {
+	sharedInformers := informers.NewSharedInformerFactory(clientset, refreshDuration)
+	controllers := make(map[string]cache.SharedIndexInformer)
+	controllers["Pod"] = sharedInformers.Core().V1().Pods().Informer()
+	controllers["ReplicationController"] = sharedInformers.Core().V1().ReplicationControllers().Informer()
+	controllers["Deployment"] = sharedInformers.Apps().V1beta1().Deployments().Informer()
+	controllers["ReplicaSet"] = sharedInformers.Apps().V1beta2().ReplicaSets().Informer()
+	controllers["StatefulSet"] = sharedInformers.Apps().V1beta2().StatefulSets().Informer()
+	controllers["Job"] = sharedInformers.Batch().V1().Jobs().Informer()
+	controllers["CronJob"] = sharedInformers.Batch().V1beta1().CronJobs().Informer()
+	controllers["Service"] = sharedInformers.Core().V1().Services().Informer()
+	controllers["Endpoints"] = sharedInformers.Core().V1().Endpoints().Informer()
+
+	return controllers
+}
+
+func (c *controllerImpl) Start() {
+	if c.stopChan == nil {
+		c.stopChan = make(chan struct{})
+		go c.run(c.stopChan)
+		log.Infof("K8S cache started")
+	} else {
+		log.Warningf("K8S cache is already running")
+	}
+}
+
+func (c *controllerImpl) run(stop <-chan struct{}) {
+	for _, cn := range c.controllers {
+		go cn.Run(stop)
+	}
+	<-stop
+	log.Infof("K8S cache stopped")
+}
+
+func (c *controllerImpl) HasSynced() bool {
+	if c.syncCount > c.maxSyncCount {
+		log.Errorf("Max attempts reached syncing cache. Error connecting to k8s API ")
```

Comment:
I think you should log the c.syncCount here. This is using Errorf and I see a blank space after "API " in the message, so maybe you forgot? I would like to see this:

`log.Errorf("Max attempts reached syncing cache. Error connecting to k8s API: %v > %v", c.syncCount, c.maxSyncCount)`